"Target Size Enhanced" - atomic rule (#2279)
auto-wcag/auto-wcag: pages/glossary/user-agent-controlled-component.md

---
title: User-Agent Controlled Component
key: user-agent-controlled-component
unambiguous: true
objective: true
input_aspects:
  - CSS styling
  - DOM tree
---

A _User-Agent Controlled Component_ is an [HTML element][namespace element] for which all the following are true:

- the element has an [implicit role][] which is a [semantic `widget`][semantic role]; and
- the [computed values][] of the element's `height` and `width` CSS properties do not depend on content provided by the author; and
- the [computed values][] of the element's `height` and `width` CSS properties do not depend on any CSS property with a [cascaded value][] with "Author" [origin][].

#### Examples of User-Agent Controlled Components

> Typically, radio buttons or checkboxes are User-Agent controlled, until an author does change their dimensions.

Reviewer: ```suggestion
Typically, radio buttons or checkboxes are User-Agent controlled, until an author changes their dimensions.
```

> Links and buttons usually aren't, because their (text) content is provided by the author and the width depends on it.

Reviewer: ```suggestion
Links and buttons usually aren't. Authors provide their (text) content which determines the width.
```

> Days in a calendar widget build with an `<input type="date">` element are also User-Agent controlled since their content isn't provided by the author and their dimension do not depend on values provided by the author.

Reviewer 2: ```suggestion
Days in a calendar widget build with an `<input type="date">` element are also User-Agent controlled since their content isn't provided by the author and their dimension does not depend on values provided by the author.
```

> Days in a calendar widget build with an `<input type="date">` element are also User-Agent controlled since their content isn't provided by the author and their dimension do not depend on values provided by the author.

Reviewer: ```suggestion
Days in a calendar widget build with an `<input type="date">` element are also User-Agent controlled. Their content is not provided by the author and dimensions do not depend on values provided by the author.
```
